File: airesources/Julia/hlt/game.jl
logger = get_logger("bot-logger")
set_level(logger, "debug")

struct Game
    id::String
    botname::String
    width::Int
    height::Int
    initial_game_map::GameMap
end
function Game(botname::String)
    id = readline()
    botname = "$botname-$id"
    width, height = parse.(Int, split(readline()))
    initial_game_map = GameMap(id, width, height, readline())
    Game(id, botname, width, height, initial_game_map)
end

GameMap(game::Game, map_string::String) = GameMap(game.id, game.width, game.height, map_string)

"""
update_map(game::Game)

Parse the map given by the engine.
"""
update_map(game::Game) = GameMap(game, readline())

"""
    start_game(game)

This function should be used after bot initialization and all preliminary checks and warmup.
"""
start_game(game::Game) = print(game.botname, "\n")

"""
    setup_logger(name, id)

Set up and truncate the log. Log name is "{id}-{name}.log".
"""
function setup_logger(name, id)
    add_handler(logger, DefaultHandler(open("$id-$name.log", "w"), 
                    DefaultFormatter("[{date} | {level} | {name}]: {msg}"))) 
    remove_handler(logger, "console")
end

"""
    send_command_queue(command_queue::Vector{String})

Issue the given list of commands.
"""
function send_command_queue(command_queue::Vector{String}) 
    print.(command_queue)
    print("\n")
end

File: airesources/Rust/src/hlt/game.rs
use std::io::{stdin, stdout, Write};
use hlt::parse::Decodable;
use hlt::entity::GameState;
use hlt::command::Command;
use hlt::game_map::GameMap;

#[derive(Debug)]
pub struct Game {
    pub my_id: usize,
    pub map_width: i32,
    pub map_height: i32,
}

impl Game {
    fn read_line() -> String {
        let mut buffer = String::new();
        stdin().read_line(&mut buffer).expect("Read error");
        buffer
    }

    fn read_id() -> usize {
        let line = Game::read_line();
        let parts = line.split_whitespace();
        let mut iter = parts.into_iter();
        usize::parse(&mut iter)
    }

    fn read_size() -> (i32, i32) {
        let line = Game::read_line();
        let parts = line.split_whitespace();
        let mut iter = parts.into_iter();
        let width = i32::parse(&mut iter);
        let height = i32::parse(&mut iter);
        (width, height)
    }

    pub fn new() -> Game {
        let my_id = Game::read_id();
        let (map_width, map_height) = Game::read_size();

        Game {
            my_id,
            map_width,
            map_height,
        }
    }

    /// Send your bot name, terminating the preprocessing
    /// time of 60 seconds allowed at the start of a game
    pub fn send_ready(&self, name: &str) {
        println!("{}", name)
    }

    /// Retrieve the new updated map
    pub fn update_map(&self) -> GameMap {
        let line = Game::read_line();
        let parts = line.split_whitespace();
        let mut iter = parts.into_iter();
        let game_state = GameState::parse(&mut iter);
        GameMap::new(self, game_state)
    }

    /// Send all commands to the game
    pub fn send_command_queue(&self, commands: &[Command]) {
        for command in commands {
            let encoded = command.encode();
            stdout().write(encoded.as_bytes()).unwrap();
        }
        println!()
    }
}
